File: app/src/main/java/net/coding/program/ImagePagerActivity.java
/**
 * ****************************************************************************
 * Copyright 2011-2013 Sergey Tarasevich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * *****************************************************************************
 */
package net.coding.program;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBar;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.nostra13.universalimageloader.core.DisplayImageOptions;
import com.nostra13.universalimageloader.core.assist.ImageScaleType;
import com.nostra13.universalimageloader.core.display.FadeInBitmapDisplayer;

import net.coding.program.common.CustomDialog;
import net.coding.program.common.ui.BackActivity;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.EActivity;
import org.androidannotations.annotations.Extra;
import org.androidannotations.annotations.OptionsItem;
import org.androidannotations.annotations.ViewById;

import java.util.ArrayList;

@EActivity(R.layout.activity_image_pager)
public class ImagePagerActivity extends BackActivity {

    private final String SAVE_INSTANCE_INDEX = "SAVE_INSTANCE_INDEX";
    DisplayImageOptions options;
    @ViewById
    ViewPager pager;
    @Extra
    int mPagerPosition;
    @Extra
    ArrayList<String> mArrayUri;
    @Extra
    boolean isPrivate;
    @Extra
    String mSingleUri;
    @Extra
    boolean needEdit;
    ImagePager adapter;

    ArrayList<String> mDelUrls = new ArrayList<>();

    TextView mMenuImagePos;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState != null) {
            mPagerPosition = savedInstanceState.getInt(SAVE_INSTANCE_INDEX, mPagerPosition);
        }
    }

    @AfterViews
    void init() {
        if (needEdit) {
            ActionBar supportActionBar = getSupportActionBar();
            supportActionBar.setIcon(android.R.color.transparent);

            View v = getLayoutInflater().inflate(R.layout.image_pager_action_custom, null);
            mMenuImagePos = (TextView) v.findViewById(R.id.imagePos);
            supportActionBar.setCustomView(v);

            supportActionBar.setDisplayShowCustomEnabled(true);
            pager.setBackgroundColor(getResources().getColor(R.color.stand_bg));

            pager.setOnPageChangeListener(new ViewPager.OnPageChangeListener() {
                @Override
                public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

                }

                @Override
                public void onPageSelected(int position) {
                    setPosDisplay(position);
                }

                @Override
                public void onPageScrollStateChanged(int state) {

                }
            });
            setPosDisplay(0);
        } else {
            getSupportActionBar().hide();
            pager.setBackgroundColor(getResources().getColor(R.color.black));
        }

        if (mSingleUri != null) {
            mArrayUri = new ArrayList<>();
            mArrayUri.add(mSingleUri);
            mPagerPosition = 0;
        }

        options = new DisplayImageOptions.Builder()
                .showImageForEmptyUri(R.drawable.ic_default_image)
                .showImageOnFail(R.drawable.ic_default_image)
                .resetViewBeforeLoading(true)
                .cacheOnDisk(true)
                .imageScaleType(ImageScaleType.EXACTLY)
                .bitmapConfig(Bitmap.Config.RGB_565)
                .considerExifParams(true)
                .displayer(new FadeInBitmapDisplayer(300))
                .build();

        if (!isPrivate) {
            initPager();
        }

    }

    private void setPosDisplay(int position) {
        String pos = String.format("%d/%d", position + 1, mArrayUri.size());
        mMenuImagePos.setText(pos);
    }

    private void initPager() {
        adapter = new ImagePager(getSupportFragmentManager());
        pager.setAdapter(adapter);
        pager.setCurrentItem(mPagerPosition);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt(SAVE_INSTANCE_INDEX, pager.getCurrentItem());
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        mPagerPosition = savedInstanceState.getInt(SAVE_INSTANCE_INDEX, mPagerPosition);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        if (needEdit) {
            MenuInflater menuInflater = getMenuInflater();
            menuInflater.inflate(R.menu.photo_pager, menu);
        }

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public void onBackPressed() {
        if (mDelUrls.isEmpty()) {
            setResult(RESULT_CANCELED);
        } else {
            Intent intent = new Intent();
            intent.putExtra("mDelUrls", mDelUrls);
            setResult(RESULT_OK, intent);
        }

        finish();
    }

    @OptionsItem
    protected final void action_del_maopao() {
        final int selectPos = pager.getCurrentItem();
        AlertDialog dialog = new AlertDialog.Builder(this)
                .setTitle("图片")
                .setMessage("确定删除？")
                .setPositiveButton("确定", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        String s = mArrayUri.remove(selectPos);
                        mDelUrls.add(s);
                        if (mArrayUri.isEmpty()) {
                            onBackPressed();
                        } else {
                            setPosDisplay(pager.getCurrentItem());
                            adapter.notifyDataSetChanged();
                        }
                    }
                })
                .setNegativeButton("取消", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {

                    }
                })
                .show();
        CustomDialog.dialogTitleLineColor(this, dialog);
    }

    class ImagePager extends FragmentPagerAdapter {

        public ImagePager(FragmentManager fm) {
            super(fm);
        }

        @Override
        public Fragment getItem(int i) {
            return ImagePagerFragment_.builder()
                    .uri(mArrayUri.get(i))
                    .customMenu(false)
                    .build();
        }

        @Override
        public Object instantiateItem(ViewGroup container, int position) {
            ImagePagerFragment fragment = (ImagePagerFragment) super.instantiateItem(container, position);
            fragment.setData(mArrayUri.get(position));
            return fragment;
        }

        @Override
        public int getItemPosition(Object object) {
            return POSITION_NONE;
        }

        @Override
        public int getCount() {
            return mArrayUri.size();
        }
    }
}

File: widget-pick-photo/src/main/java/net/coding/program/pickphoto/detail/ImagePagerActivity.kt
/**
 * ****************************************************************************
 * Copyright 2011-2013 Sergey Tarasevich
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * *****************************************************************************
 */
package net.coding.program.pickphoto.detail

import android.app.Activity
import android.content.Intent
import android.graphics.Bitmap
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentStatePagerAdapter
import android.support.v4.view.PagerAdapter
import android.support.v4.view.ViewPager
import android.support.v7.app.AlertDialog
import android.view.Menu
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.nostra13.universalimageloader.core.DisplayImageOptions
import com.nostra13.universalimageloader.core.assist.ImageScaleType
import com.nostra13.universalimageloader.core.display.FadeInBitmapDisplayer
import net.coding.program.common.ui.BackActivity
import net.coding.program.common.util.PermissionUtil
import net.coding.program.pickphoto.R
import net.coding.program.pickphoto.R2
import org.androidannotations.annotations.*
import java.util.*

@EActivity(R2.layout.activity_image_pager)
open class ImagePagerActivity : BackActivity() {

    private val SAVE_INSTANCE_INDEX = "SAVE_INSTANCE_INDEX"
    internal lateinit var options: DisplayImageOptions

    @Extra
    @JvmField
    protected final var mPagerPosition: Int = 0
    @Extra
    @JvmField
    protected final var mArrayUri = ArrayList<String>()
    @Extra
    @JvmField
    protected final var isPrivate: Boolean = false
    @Extra
    @JvmField
    protected final var mSingleUri: String? = null
    @Extra
    @JvmField
    protected final var needEdit: Boolean = false

    @ViewById(R2.id.pager)
    protected lateinit var pager: ViewPager

    internal lateinit var adapter: ImagePager

    internal var mDelUrls = ArrayList<String>()

    internal lateinit var mMenuImagePos: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (savedInstanceState != null) {
            mPagerPosition = savedInstanceState.getInt(SAVE_INSTANCE_INDEX, mPagerPosition)
        }
    }

    @AfterViews
    protected fun initImagePagerActivity() {
        if (needEdit) {
            val supportActionBar = supportActionBar
            supportActionBar!!.setIcon(android.R.color.transparent)

            val v = layoutInflater.inflate(R.layout.image_pager_action_custom, null)
            mMenuImagePos = v.findViewById<View>(R.id.imagePos) as TextView
            supportActionBar.customView = v

            supportActionBar.setDisplayShowCustomEnabled(true)
            pager!!.setBackgroundColor(resources.getColor(R.color.stand_bg))

            pager!!.setOnPageChangeListener(object : ViewPager.OnPageChangeListener {
                override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {

                }

                override fun onPageSelected(position: Int) {
                    setPosDisplay(position)
                }

                override fun onPageScrollStateChanged(state: Int) {

                }
            })
            setPosDisplay(0)
        } else {
            supportActionBar!!.hide()
            pager!!.setBackgroundColor(resources.getColor(R.color.black))
        }

        if (mSingleUri != null) {
            mArrayUri = ArrayList()
            mArrayUri.add(mSingleUri!!)
            mPagerPosition = 0
        }

        options = DisplayImageOptions.Builder()
                .showImageForEmptyUri(R.drawable.ic_default_image)
                .showImageOnFail(R.drawable.ic_default_image)
                .resetViewBeforeLoading(true)
                .cacheOnDisk(true)
                .imageScaleType(ImageScaleType.EXACTLY)
                .bitmapConfig(Bitmap.Config.RGB_565)
                .considerExifParams(true)
                .displayer(FadeInBitmapDisplayer(300))
                .build()

        if (!isPrivate) {
            initPager()
        }

    }

    private fun setPosDisplay(position: Int) {
        val pos = String.format("%s/%s", position + 1, mArrayUri.size)
        mMenuImagePos.text = pos
    }

    private fun initPager() {
        adapter = ImagePager(supportFragmentManager)
        pager!!.adapter = adapter
        pager!!.currentItem = mPagerPosition
    }

    public override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        outState!!.putInt(SAVE_INSTANCE_INDEX, pager!!.currentItem)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        mPagerPosition = savedInstanceState.getInt(SAVE_INSTANCE_INDEX, mPagerPosition)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        if (needEdit) {
            val menuInflater = menuInflater
            menuInflater.inflate(R.menu.photo_pager, menu)
        }

        return super.onCreateOptionsMenu(menu)
    }

    override fun onBackPressed() {
        if (mDelUrls.isEmpty()) {
            setResult(Activity.RESULT_CANCELED)
        } else {
            val intent = Intent()
            intent.putExtra("mDelUrls", mDelUrls)
            setResult(Activity.RESULT_OK, intent)
        }

        finish()
    }

    @OptionsItem(R2.id.action_del_maopao)
    protected fun actionRelMaopao() {
        val selectPos = pager!!.currentItem
        val dialog = AlertDialog.Builder(this, R.style.MyAlertDialogStyle)
                .setTitle("图片")
                .setMessage("确定删除？")
                .setPositiveButton("确定") { dialog, which ->
                    val s = mArrayUri.removeAt(selectPos)
                    mDelUrls.add(s)
                    if (mArrayUri.isEmpty()) {
                        onBackPressed()
                    } else {
                        setPosDisplay(pager!!.currentItem)
                        adapter.notifyDataSetChanged()
                    }
                }
                .setNegativeButton("取消") { dialog, which -> }
                .show()
    }

    internal inner class ImagePager(fm: FragmentManager) : FragmentStatePagerAdapter(fm) {

        override fun getItem(i: Int): Fragment {
            return ImagePagerFragment_.builder()
                    .uri(mArrayUri[i])
                    .customMenu(false)
                    .build()
        }

        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            val fragment = super.instantiateItem(container, position) as ImagePagerFragment
            fragment.setData(mArrayUri[position])
            return fragment
        }

        override fun getItemPosition(`object`: Any): Int {
            return PagerAdapter.POSITION_NONE
        }

        override fun getCount(): Int {
            return mArrayUri.size
        }
    }
}